crates/net/common/src/metered_stream.rs

```diff
@@ -0,0 +1,414 @@
+        self.meters.insert(meter_name, (meter, None));
+    }
+
+    /// Attaches the provided  [`MeteredStreamMetrics`]
+    pub fn set_metrics(
+        &mut self,
+        meter_name: &'a str,
+        metrics: MeteredStreamMetrics,
+    ) -> Result<(), MeterNotFoundError> {
+        let meter_metrics = self.meters.get_mut(meter_name).ok_or(MeterNotFoundError)?;
+        meter_metrics.1 = Some(metrics);
+        Ok(())
+    }
+
+    /// Provides a reference to the [`MeteredStreamCounts`] attached to this [`MeteredStream`]
+    pub fn get_metered_stream_counts(&self, meter_name: &'a str) -> Option<&MeteredStreamCounts> {
+        self.meters.get(meter_name).map(|m| &m.0)
+    }
+}
+
+impl<'a, S> AsMut<MeteredStream<'a, S>> for MeteredStream<'a, S> {
+    fn as_mut(&mut self) -> &mut MeteredStream<'a, S> {
+        self
+    }
+}
+
+impl<Stream: AsyncRead> AsyncRead for MeteredStream<'_, Stream> {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        let this = self.project();
+        let num_bytes_u64 = {
+            let init_num_bytes = buf.filled().len();
+            ready!(this.inner.poll_read(cx, buf))?;
+            u64::try_from(buf.filled().len() - init_num_bytes).unwrap_or(u64::max_value())
+        };
+
+        this.meters.iter().for_each(|(_, (meter, metrics))| {
+            let current_ingress =
+                meter.inner.ingress.fetch_add(num_bytes_u64, Ordering::Relaxed) + num_bytes_u64;
+
+            if let Some(metered_stream_metrics) = metrics {
+                metered_stream_metrics.inner.ingress_bytes.absolute(current_ingress);
+            }
+        });
+
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl<Stream: AsyncWrite> AsyncWrite for MeteredStream<'_, Stream> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.project();
+        let num_bytes = ready!(this.inner.poll_write(cx, buf))?;
+        let num_bytes_u64 = { u64::try_from(num_bytes).unwrap_or(u64::max_value()) };
+
+        this.meters.iter().for_each(|(_, (meter, metrics))| {
+            let current_egress =
+                meter.inner.egress.fetch_add(num_bytes_u64, Ordering::Relaxed) + num_bytes_u64;
+
+            if let Some(metered_stream_metrics) = metrics {
+                metered_stream_metrics.inner.egress_bytes.absolute(current_egress);
+            }
+        });
+
+        Poll::Ready(Ok(num_bytes))
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let this = self.project();
+        this.inner.poll_flush(cx)
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let this = self.project();
+        this.inner.poll_shutdown(cx)
+    }
+}
+
+impl MeteredStream<'_, UdpSocket> {
```

Comment:
hmm it is a bit odd that we implement something called Stream for a UdpSocket

since it only sends individual packets, perhaps counting bytes read/written is not really ideal